### ISSUE TITLE
Runtime TypeError when reusing variables on left side of let-expressions.

### ISSUE THREAD

**Issue by drathier** [elm-lang/elm-lang.org#647]
I managed to write a test-case that triggered a Javascript error. I think it has to do with shadowing variables from outer scope in `let` expressions. I'm using Elm 0.17.1 on Win10 64bit.

This is the code that found the error. 
```
fuzz2 Fuzz.int Fuzz.int "mvce" <|
    \a b ->
        case Crdt.empty keySz segSz 42 of 
          Nothing -> Expect.fail "Crdt.empty returned Nothing"
          Just crdt ->
            let
              crdt = Crdt.add a <| Crdt.add b <| crdt
              values = Set.fromList <| Dict.values <| Crdt.fetchAll crdt
            in Expect.equal values <| Set.fromList [a,b]
```
where the functions have the following types:
```
module Crdt exposing
  ( Crdt
  , empty
  , add
  , remove
  , fetchAll
  , merge
  , seen
  )

empty : Int -> Int -> Int -> Maybe (Crdt dataType)
add : dataType -> Crdt dataType -> Crdt dataType
fetchAll : Crdt dataType -> Dict.Dict Int dataType

type Crdt dataType = Crdt (IntCrdt dataType)
```

The `IntCrdt dataType` record holds `Dict Int dataType`, `Set Int`, `Int`, `Random.Seed`.

Renaming the inner `crdt` variable fixes the problem.
```
, fuzz2 Fuzz.int Fuzz.int "mvce" <|
    \a b ->
        case Crdt.empty keySz segSz 42 of
          Nothing -> Expect.fail "Crdt.empty returned Nothing"
          Just crdt ->
            let
              ncrdt = Crdt.add a <| Crdt.add b <| crdt
              values = Set.fromList <| Dict.values <| Crdt.fetchAll ncrdt
            in Expect.equal values <| Set.fromList [a,b]
```

Here's the TypeError:
```
$ elm-test
Success! Compiled 1 module.
Successfully generated C:\Users\drathier\AppData\Local\Temp\elm_test_116920-15072-c1in4g.j5jwqm2t9.js

elm-test
--------

Running 7 tests. To reproduce these results, run: elm-test --seed 1413460212

undefined:13294
                var _p11 = _p10._0;
                               ^

TypeError: Cannot read property '_0' of undefined
    at Function.eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:13294:18)
    at A2 (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:92:11)
    at Function.eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:13470:10)
    at A2 (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:92:11)
    at eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:765:10)
    at eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:7:53)
    at eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:10332:18)
    at Function.eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:10342:7)
    at A3 (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:98:11)
    at Function.eval (eval at <anonymous> (C:\Users\drathier\AppData\Roaming\npm\node_modules\elm-test\bin\elm-test:86:37), <anonymous>:10358:20)
```

Related errors, found by trying to create a sscce/mvce:
```
fuzz Fuzz.int "mvce" <|
  \a ->
    case Just a of
      Nothing -> Expect.fail "wat"
      Just x ->
        let
          x = x + 1
          values = x + 1
        in Expect.equal values (a+2)
```
finds the testcase `x=0` where `2 != NaN`. Adding integers together gives `NaN`.

**Comment by drathier** [comment]
If you really need the full source code to recreate this, I can invite you to the private Github repo.

**Comment by evancz** [comment]
Thanks for the report.

First, can you reopen this [here](https://github.com/elm-lang/elm-compiler/issues) instead of on the website repo? Before opening, is it the same as https://github.com/elm-lang/elm-compiler/issues/873? If so, is this fixed in the 0.18 alpha?

If not, can you get it down to an example I can actually run? It is very hard to debug issues like this without having code to experiment with.
